### PR TITLE
Add Objective-C language support

### PR DESCRIPTION
**Pull request by steinuil**
I only pulled the tree-sitter grammar from https://github.com/jiyee/tree-sitter-objc/ and pasted the highlight.scm file from that repository, so there might be things to improve.

### CODE REVIEW

File: runtime/queries/objc/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+  "volatile"
+  "goto"
+  "register"
+  "__covariant"
+  "__contravariant"
+  "__GENERICS"
+] @keyword
+
+"sizeof" @keyword.operator
+"return" @keyword.return
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:88]:
```suggestion
"return" @keyword.control.return
```

File: runtime/queries/objc/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+[
+  "while"
+  "for"
+  "do"
+  "continue"
+  "break"
+] @keyword.repeat
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:96]:
```suggestion
] @keyword.control.repeat
```

File: runtime/queries/objc/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+[
+  "#if"
+  "#ifdef"
+  "#ifndef"
+  "#else"
+  "#elif"
+  "#endif"
+  (preproc_directive)
+] @keyword
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:108]:
```suggestion
"#define" @keyword.directive

[
  "#if"
  "#ifdef"
  "#ifndef"
  "#else"
  "#elif"
  "#endif"
  (preproc_directive)
] @keyword.directive
```

File: runtime/queries/objc/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+"#include" @include
+"#import" @include
+"@import" @include
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:112]:
```suggestion
"#include" @keyword.include
"#import" @keyword.include
"@import" @keyword.include
```

File: runtime/queries/objc/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+] @operator
+
+[
+ "if"
+ "else"
+ "case"
+ "switch"
+] @keyword.conditional
+
+(conditional_expression [ "?" ":" ] @keyword.conditional)
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:165]:
```suggestion
] @keyword.control.conditional

(conditional_expression [ "?" ":" ] @keyword.control.conditional)
```

File: runtime/queries/objc/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+[
+ (true)
+ (false)
+ (YES)
+ (NO)
+] @keyword.boolean
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:172]:
```suggestion
] @constant.builtin.boolean
```

File: runtime/queries/objc/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+[ "(" ")" "[" "]" "{" "}"] @punctuation.bracket
+
+[
+  (string_literal)
+  (string_expression)
+  (system_lib_string)
+  (module_string)
+] @string
+
+(escape_sequence) @string.escape
```
Comment (the-mikedavis) [inline runtime/queries/objc/highlights.scm:187]:
```suggestion
(escape_sequence) @constant.character.escape
```

**Review by the-mikedavis — COMMENTED** [review]
We use some custom captures documented here: https://docs.helix-editor.com/master/themes.html#syntax-highlighting

I've left some example changes as review comments but there are more below L190 that should be changed to use the Helix captures instead too

**Comment by kasperisager** [comment]
This would be a fantastic addition! @steinuil, do you need a hand with any of this?

**Comment by steinuil** [comment]
Hi @kasperisager, yes gladly! I know very little Objective-C so I had a hard time figuring out how to categorize the keywords, I'll push what I have so if you want you can start with that.

**Comment by kasperisager** [comment]
@steinuil I've just committed and pushed the grammars and queries I've been using for Obj-C and Obj-C++, also based on https://github.com/jiyee/tree-sitter-objc: https://github.com/kasperisager/dotfiles/commit/0193bb537323a4fe40feb379f19e2d54beb81794.

@the-mikedavis I initially tried making the Obj-C queries inherit from the plain C queries, which unfortunately didn't work as Helix choked on certain queries, such as `(attribute name: (identifier) @attribute)`. Could this be because the Obj-C tree sitter parser doesn't recognise `attribute` nodes?

**Comment by pascalkuthe** [comment]
grammer are not usually related to each other and just using queries from one grammar for another unrelated language doesn't work

**Comment by pascalkuthe** [comment]
I am closing this PR PR as stale. Te captures were just copied from upstream and not adapted to helix. If somebody wants to add objective-c language support to helix they should do that and open a new PR. Thank you for contributing

**Pull request closed without merging**